go-controller/pkg/ovn/policy.go

```diff
@@ -1071,80 +1078,88 @@ func (oc *Controller) createNetworkPolicy(np *networkPolicy, policy *knet.Networ
 		podSelector       *metav1.LabelSelector
 	}
 	var policyHandlers []policyHandler
-	// Go through each ingress rule.  For each ingress rule, create an
-	// addressSet for the peer pods.
-	for i, ingressJSON := range policy.Spec.Ingress {
-		klog.V(5).Infof("Network policy ingress is %+v", ingressJSON)
 
-		ingress := newGressPolicy(knet.PolicyTypeIngress, i, policy.Namespace, policy.Name)
+	policyTypeIngress, policyTypeEgress := policyType(policy)
```

Comment:
is this additional check really needed? If policy.Spec.Ingress not empty, ingress should be set to true right?

Reply:
I mean it's kubeapi job to make sure policyType is correctly set, but ovn-k should work based on NetworkPolicySpec, and it says 
```
ingress NetworkPolicyIngressRule array | List of ingress rules to be applied to the selected pods. ...
If this field is empty then this NetworkPolicy does not allow any traffic (and serves solely to ensure that the pods it selects are isolated by default)
```
https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.19/#networkpolicyspec-v1-networking-k8s-io
I don't see why we need to check policyTypes before applying NetworkPolicyIngressRules

Reply:
If policyTypes is set to egress only, the policy does not express any isolation for ingress.

This is how you default deny ingress traffic:
https://kubernetes.io/docs/concepts/services-networking/network-policies/#default-deny-all-ingress-traffic

Notice how a default deny for egress is not expressing any isolation for ingress:
https://kubernetes.io/docs/concepts/services-networking/network-policies/#default-deny-all-egress-traffic
```
This policy does not change the ingress isolation behavior of any pod.
```

Why would we not check policyTypes always if it is always explicitly expressing the isolation intention of the policy, regardless of whether is a default deny policy or not?
I think that the documentation you are quoting is implicitly conditioned to what policyTypes indicates. It says the same thing for egress rules.

Reply:
isolation is defined in another part of the code, where ports are added to default deny port group, and it is done based on policytype. i was talking about adding allow ACLs, which is done based on `policy.Spec.Ingress`. i just thought that if `policy.Spec.Ingress` not empty, policytype will always have ingress in it, so this check will be redundant.
But having this if doesn't make anything worse anyway.

Reply:
Hmm now that I think of it, it actually does make things worse.

We need to add the policy ACLs even if a pod is not isolated, because if it becomes isolated through another policy added at a later time, we are not going back to process all network policies rules.
That means that the ACLs we add when the pod is not isolated are actually noop, and not adding them in this case would be an improvement, but this could be a performance hit.

The purpose of this PR is if a pod is selected by a single policy, for example
```
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: test-network-policy
  namespace: default
spec:
  podSelector:
    matchLabels:
      role: db
  policyTypes:
    - Egress
  ingress:
    - from:
        - ipBlock:
            cidr: 172.17.0.0/16
            except:
              - 172.17.1.0/24
        - namespaceSelector:
            matchLabels:
              project: myproject
        - podSelector:
            matchLabels:
              role: frontend
      ports:
        - protocol: TCP
          port: 6379
  egress:
    - to:
        - ipBlock:
            cidr: 10.0.0.0/24
      ports:
        - protocol: TCP
          port: 5978
```
That it does not become isolated for ingress because even it has ingress rules, the policy type is not ingress.